src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -331,78 +342,163 @@ private[concurrent] object Promise {
+        val executor = a.asInstanceOf[ExecutionContext]
+        _arg = v
+        try executor.execute(this) catch { case NonFatal(t) => executor reportFailure t }
+      }
+    }
+
+    // Gets invoked by run(), runs the transformation and stores the result
+    private[this] final def transform(v: Try[F], fn: Try[F] => M[T]): Unit = {
+      val res = fn(v)
+      if (res.isInstanceOf[Try[T @unchecked]]) this complete res.asInstanceOf[Try[T]]
+      else if (res.isInstanceOf[DefaultPromise[T @unchecked]]) res.asInstanceOf[DefaultPromise[T]].linkRootOf(this)
+      else if (res.isInstanceOf[Future[T @unchecked]]) this completeWith res.asInstanceOf[Future[T]]
+      else this success res.asInstanceOf[T]
+    }
+
+    // Gets invoked by the ExecutionContext, when we have a value to transform
+    override final def run(): Unit = {
+      val v = _arg
+      _arg = null
+      if (v.isInstanceOf[Try[F]]) {
+        val fun = _fun
+        _fun = null
+        try transform(v.asInstanceOf[Try[F]], fun) catch {
+          case NonFatal(t) => this failure t
+        }
+      }
+    }
 
-    private[this] sealed trait Kept[T] extends Promise[T] {
-      def result: Try[T]
+    override final def toString: String = super[DefaultPromise].toString
+  }
+
+  /* Encodes the concept of having callbacks.
+   * This is an `abstract class` to make sure calls are `invokevirtual` rather than `invokeinterface`
+   */
+  sealed trait Callbacks[+T] {
+    /* Logically prepends the callback `c` onto `this` callback */
+    def prepend[U >: T](c: Callbacks[U]): Callbacks[U]
+    /* Submits the callback function(s) represented by this Callback to be executed with the given value `v` */
+    def submitWithValue(v: Try[T @uncheckedVariance]): Unit
+  }
 
-      override def value: Option[Try[T]] = Some(result)
+  /* Represents 0 Callbacks, is used as an initial, sentinel, value for DefaultPromise
+   * This used to be a `case object` but in order to keep `Callbacks`'s methods bimorphic it was reencoded as `val`
+   */
+  final object NoopCallback extends Callbacks[Nothing] {
+    override def prepend[U >: Nothing](c: Callbacks[U]): Callbacks[U] = c 
+    override def submitWithValue(v: Try[Nothing @uncheckedVariance]): Unit = ()
+  }
 
-      override def isCompleted: Boolean = true
+  final object ManyCallbacks {
+    final def two[U](second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](NoopCallback, NoopCallback, second, first, 2)
+    final def three[U](third: Callbacks[U], second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](NoopCallback, third, second, first, 1)
+    final def four[U](fourth: Callbacks[U], third: Callbacks[U], second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](fourth, third, second, first, 0)
+  }
 
-      override def tryComplete(value: Try[T]): Boolean = false
+  final class ManyCallbacks[+T] private[ManyCallbacks] (
```

Comment:
I've tried to use something similar but, if my memory is correct, having multiple fields didn't improve throughput and increased the memory footprint. For both traneio and twitter futures, we used a simple queue similar to a linked list:

https://github.com/traneio/future/blob/master/future-java/src/main/java/io/trane/future/WaitQueue.java#L12
https://github.com/twitter/util/blob/develop/util-core/src/main/scala/com/twitter/util/Promise.scala#L57

Reply:
@fwbrasil My assumption is that most promises either have 0, 1, or many callbacks, so optimizing for 2…3 callbacks didn't make much sense. (The old Queueing Theory adage that a queue spends most of its life either full or empty).

With TransformationalPromise, there is no extra allocation for a Promise with a single, known-beforehand, callback, and it only adds the cost of 2 fields to a class which already has only 1 field. I guess it can be debated whether a ManyCallbacks-breadth of 2, 3, or 4 makes the most sense from a memory pressure perspective, but one has to remember the 10byte object header overhead, which incidentally is 1…2 references (32-bit refs or 64-bit refs).

also, since there is less pointer-chasing there is less cache-pollution. :)

Reply:
> @fwbrasil My assumption is that most promises either have 0, 1, or many callbacks, so optimizing for 2…3 callbacks didn't make much sense. (The old Queueing Theory adage that a queue spends most of its life either full or empty).

I think it depends a lot on how the futures are used. For instance, it's common to see a pattern at Twitter where there's a callback to update metrics + the regular call for future composition. I imagine that's why having a linked list is more efficient for us since it introduces less overhead for queues with two elements.